src/lib.rs

```diff
@@ -50,36 +55,48 @@ pub fn run(args: Args) {
     };
     let supported_language = args.language.get_language();
     let language = supported_language.language();
-    let query = get_query(&query_source, language);
+    let query = Arc::new(get_query(&query_source, language));
     let capture_index = args.capture_name.as_ref().map_or(0, |capture_name| {
         query
             .capture_index_for_name(capture_name)
             .expect(&format!("Unknown capture name: `{}`", capture_name))
     });
 
-    enumerate_project_files(&*supported_language)
-        .par_iter()
-        .for_each(|project_file_dir_entry| {
-            let mut printer = grep::printer::Standard::new_no_color(io::stdout());
-            let path = project_file_dir_entry.path();
-
-            let matcher = TreeSitterMatcher::new(
-                &query,
-                capture_index,
-                language,
-                args.filter.clone(),
-                args.filter_arg.clone(),
-            );
-
-            SearcherBuilder::new()
-                .multi_line(true)
-                .build()
-                .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
-                .unwrap();
-        })
+    let (sender, receiver) = mpsc::channel::<DirEntry>();
+    let handle = thread::spawn(move || {
+        let thread_pool = ThreadPool::new(num_cpus::get());
+        while let Ok(project_file_dir_entry) = receiver.recv() {
+            thread_pool.execute({
+                let query = query.clone();
+                let args_filter = args.filter.clone();
+                let args_filter_arg = args.filter_arg.clone();
+                move || {
+                    let mut printer = grep::printer::Standard::new_no_color(io::stdout());
+                    let path = project_file_dir_entry.path();
+
+                    let matcher = TreeSitterMatcher::new(
+                        &query,
+                        capture_index,
+                        language,
+                        args_filter,
+                        args_filter_arg,
+                    );
+
+                    SearcherBuilder::new()
+                        .multi_line(true)
+                        .build()
+                        .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
+                        .unwrap();
+                }
+            });
+        }
+        thread_pool.join();
```

Comment:
This blocks on all of the thread pool jobs finishing so therefore the "receiver" thread shouldn't return until (a) the project file enumeration has completed (at which point the sender should get dropped and `receiver.recv()` should then return an `Err`) and (b) all of the thread-pool query-matching jobs have run to completion